modyn/supervisor/internal/pipeline_executor/evaluation_executor.py

```diff
@@ -0,0 +1,331 @@
+"""Coordinates the evaluation after the core pipeline execution."""
+
+import datetime
+import logging
+import pickle
+import sys
+from concurrent.futures import Future, ThreadPoolExecutor
+from dataclasses import dataclass
+from functools import partial
+from multiprocessing import Queue
+from pathlib import Path
+
+import pandas as pd
+from modyn.config.schema.pipeline import ModynPipelineConfig
+from modyn.config.schema.system import ModynConfig
+
+# pylint: disable-next=no-name-in-module
+from modyn.evaluator.internal.grpc.generated.evaluator_pb2 import EvaluateModelResponse, EvaluationAbortedReason
+from modyn.supervisor.internal.eval.handler import EvalHandler, EvalRequest
+from modyn.supervisor.internal.eval.result_writer.json_result_writer import JsonResultWriter
+from modyn.supervisor.internal.grpc.enums import IdType, MsgType, PipelineStage
+from modyn.supervisor.internal.grpc.template_msg import id_submsg, pipeline_stage_msg
+from modyn.supervisor.internal.grpc_handler import GRPCHandler
+from modyn.supervisor.internal.pipeline_executor.models import (
+    ConfigLogs,
+    PipelineLogs,
+    SingleEvaluationInfo,
+    StageLog,
+    SupervisorLogs,
+)
+from modyn.supervisor.internal.utils.evaluation_status_reporter import EvaluationStatusReporter
+from modyn.utils.utils import current_time_micros, dynamic_module_import
+from pydantic import BaseModel
+
+eval_strategy_module = dynamic_module_import("modyn.supervisor.internal.eval.strategies")
+
+logger = logging.getLogger(__name__)
+
+
+class EvalStateConfig(BaseModel):
+    pipeline_id: int
+    eval_dir: Path
+    config: ModynConfig
+    pipeline: ModynPipelineConfig
+
+
+@dataclass
+class AfterPipelineEvalContext:
+    tracking_dfs: dict[str, pd.DataFrame]
+
+
+class EvaluationExecutor:
+    def __init__(
+        self,
+        pipeline_id: int,
+        pipeline_logdir: Path,
+        config: ModynConfig,
+        pipeline: ModynPipelineConfig,
+        grpc: GRPCHandler,
+    ):
+        self.pipeline_id = pipeline_id
+        self.pipeline_logdir = pipeline_logdir
+        self.config = config
+        self.pipeline = pipeline
+        self.grpc = grpc
+        self.context: AfterPipelineEvalContext | None = None
+        self.eval_handlers = (
+            [EvalHandler(eval_handler_config) for eval_handler_config in pipeline.evaluation.handlers]
+            if pipeline.evaluation
+            else []
+        )
+
+    def register_tracking_info(self, tracking_dfs: dict[str, pd.DataFrame]) -> None:
+        assert tracking_dfs.get(PipelineStage.HANDLE_SINGLE_TRIGGER.name) is not None
+        assert tracking_dfs.get(PipelineStage.STORE_TRAINED_MODEL.name) is not None
+        self.context = AfterPipelineEvalContext(tracking_dfs=tracking_dfs)
+
+    def create_snapshot(self) -> None:
+        """Create a snapshot of the evaluation state before starting to evaluate."""
```

Comment:
I guess this is more like training/pipeline metadata? Evaluation state to me sounds  very much like we snapshot the state of the evaluation (partial progress).

Reply:
agreed